src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -98,17 +105,53 @@ def run_python_test(transitive_hydrated_target, pytest, python_setup):
   requirements_pex_response = yield Get(
     ExecuteProcessResult, ExecuteProcessRequest, requirements_pex_request)
 
-  # Gather sources.
+  source_roots = source_root_config.get_source_roots()
+
+  # Gather sources and adjust for the source root.
   # TODO: make TargetAdaptor return a 'sources' field with an empty snapshot instead of raising to
   # simplify the hasattr() checks here!
+  # TODO: find some way to restore the full source name for the stdout of the Pytest run, so that
+  # we can output the original path, rather than the path without the source root. It appears we
+  # do this in V1 via parsing the Junit XML file and adding the value back? See pytest_run.py's
+  # _get_target_from_test(). This is a low priority.
   all_sources_digests = []
   for maybe_source_target in all_targets:
     if hasattr(maybe_source_target.adaptor, 'sources'):
       sources_snapshot = maybe_source_target.adaptor.sources.snapshot
-      all_sources_digests.append(sources_snapshot.directory_digest)
+      digest_adjusted_for_source_root = yield Get(
+        Digest,
+        DirectoryWithPrefixToStrip(
+          sources_snapshot.directory_digest,
+          source_roots.find_by_path(maybe_source_target.adaptor.address.spec_path).path
+        )
+      )
+      all_sources_digests.append(digest_adjusted_for_source_root)
+
+  sources_digest = yield Get(
+    Digest, MergedDirectories(directories=tuple(all_sources_digests)),
+  )
+
+  # TODO: add intrinsic to go from Digest->Snapshot, so that we can avoid having to use
+  # `FileContent`, which unnecessarily gets the content.
```

Comment:
I think we also might be able to make whatever ruleset feeds `MergedDirectories->Digest` to instead return a `Snapshot`? Whatever rule is creating the `Digest` probably has the path stats available too (maybe). But that might not be a good use of time if the API is easier to allows `Digest->Snapshot` because digests are easier to pass around for some reason.

Still, I might modify the TODO to maybe get a `Snapshot` from the `MergedDirectories` if possible (likely with some Rust changes), but otherwise `Digest->Snapshot`.

And ideally all TODOs always have issue links!! Not everyone does it for some reason, but that's the idea, unless the TODO is extremely focused to the specific region of code it's in, imho?